atomix/cluster/src/main/java/io/atomix/raft/roles/LeaderRole.java

```diff
@@ -475,63 +469,53 @@ public CompletableFuture<VoteResponse> onVote(final VoteRequest request) {
     }
   }
 
-  /**
-   * Appends an entry to the Raft log.
-   *
-   * @param entry the entry to append
-   * @return a completable future to be completed once the entry has been appended
-   */
-  private CompletableFuture<IndexedRaftLogEntry> append(final RaftLogEntry entry) {
-    CompletableFuture<IndexedRaftLogEntry> resultingFuture = null;
+  private IndexedRaftLogEntry appendEntry(final RaftLogEntry entry) {
+    try {
+      return appendWithRetry(entry);
+    } catch (final Exception e) {
+      log.error("Failed to append to local log, stepping down", e);
+      raft.transition(Role.FOLLOWER);
+      throw e;
+    }
+  }
+
+  private IndexedRaftLogEntry appendWithRetry(final RaftLogEntry entry) {
     int retries = 0;
 
-    do {
+    RuntimeException lastError = null;
+    // we retry in a blocking fashion to avoid interleaving append requests; this however blocks the
+    // raft thread.
+    while (retries <= MAX_APPEND_ATTEMPTS) {
       try {
-        resultingFuture = tryToAppend(entry);
-      } catch (final JournalException storageException) {
+        return append(entry);
+      } catch (final JournalException.OutOfDiskSpace e) {
+        // if this happens then compact will also not help, since we need to create a snapshot
```

Comment:
That's probably true, since we used to take snapshots from the same context. I can add calling compact here and then retrying, and immediately increasing the retry count to the max (so we only retry once)?